clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/safety/TrafficGuard.java

```diff
@@ -0,0 +1,377 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package com.netflix.spinnaker.clouddriver.safety;
+
+import static java.lang.String.format;
+
+import com.google.common.collect.ImmutableMap;
+import com.netflix.spectator.api.Id;
+import com.netflix.spectator.api.Registry;
+import com.netflix.spectator.impl.Preconditions;
+import com.netflix.spinnaker.clouddriver.core.services.Front50Service;
+import com.netflix.spinnaker.clouddriver.exceptions.TrafficGuardException;
+import com.netflix.spinnaker.clouddriver.model.Cluster;
+import com.netflix.spinnaker.clouddriver.model.ClusterProvider;
+import com.netflix.spinnaker.clouddriver.model.HealthState;
+import com.netflix.spinnaker.clouddriver.model.ServerGroup;
+import com.netflix.spinnaker.clouddriver.names.NamerRegistry;
+import com.netflix.spinnaker.kork.dynamicconfig.DynamicConfigService;
+import com.netflix.spinnaker.moniker.Moniker;
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import retrofit.RetrofitError;
+
+@Component
+public class TrafficGuard {
+  private static final String MIN_CAPACITY_RATIO = "traffic-guards.min-capacity-ratio";
+  private final Logger log = LoggerFactory.getLogger(getClass());
+
+  private final List<ClusterProvider<?>> clusterProviders;
+  private final Front50Service front50Service;
+  private final Registry registry;
+  private final DynamicConfigService dynamicConfigService;
+
+  private final Id savesId;
+
+  @Autowired
+  public TrafficGuard(
+      List<ClusterProvider<?>> clusterProviders,
+      Optional<Front50Service> front50Service,
+      Registry registry,
+      DynamicConfigService dynamicConfigService) {
+    this.clusterProviders = clusterProviders;
+    this.front50Service = front50Service.orElse(null);
+    this.registry = registry;
+    this.dynamicConfigService = dynamicConfigService;
+    this.savesId = registry.createId("trafficGuard.saves");
+  }
+
+  public void verifyInstanceTermination(
+      String serverGroupName,
+      List<String> instanceIds,
+      String account,
+      String location,
+      String cloudProvider,
+      String operationDescriptor) {
+    // TODO(rz): I opted out of migrating this method because it isn't used in
+    //  my current refactors. This method uses clouddriver search endpoint,
+    //  which would be a much larger refactor to bring over in this commit. I
+    //  would like to postpone such a refactor until it's actually needed.
```

Comment:
How long will this class be both in `clouddriver` and `orca`?  Given that it looks from this comment like there are still usages in `orca`, is the plan for this to live in both places indefinitely, or is there a plan to eventually move the orca usages here?

Reply:
I guess one question would be could we (in some semi short-term) move everything over and just have a shim in `orca` that makes an API call to a traffic guard api backed by this class. Perhaps that's what you mean by move the `orca` usages here @ezimanyi.

We're already making a call across the wire and at least if we can get the business logic all on one side we wouldn't have to worry so much about drift.

I like this logic being in `clouddriver` and see it moving across as a win.

Reply:
Introducing an API is probably the right move - I have no idea how long it will take to migrate all cloud provider logic into Orca -- probably years, unless cloud providers jump on the bandwagon to help the migration process.

It'd be nice to have help on this effort. I'll add creating a Traffic Guard API to clouddriver to my backlog, but getting the initial cloud operation stage in is my higher priority.